go/private/actions/archive.bzl

```diff
@@ -65,10 +67,26 @@ def emit_archive(go, source = None, _recompile_suffix = ""):
     direct = [get_archive(dep) for dep in source.deps]
     runfiles = source.runfiles
     data_files = runfiles.files
-    for a in direct:
-        runfiles = runfiles.merge(a.runfiles)
-        if a.source.mode != go.mode:
-            fail("Archive mode does not match {} is {} expected {}".format(a.data.label, mode_string(a.source.mode), mode_string(go.mode)))
+
+    # After bazel version 5.0.0, skylark introduced a new API called merge_all().
+    # There is a recommendation in the release notes to use merge_all() if
+    # we hit the depth limit because of using merge() in a loop, we saw
+    # some repos hitting this issue after upgrading to 5.0.0
+    # To keep it backwards compatible, we will fallback to the old logic if
+    # the version is not at least 5.0.0
```

Comment:
nit: you can also add a TODO saying that we can remove this check after the minimal version of Bazel becomes 5.0.0, which we will have after integrating with bzlmod